### ISSUE TITLE
ED2 + infiniband

### ISSUE THREAD

**Issue by serbinsh** [EDmodel/ED2#167]
All

Here is an issue I am getting on our cluster which we recently upgraded and added infiniband.  When compiled with openMPI and when we have infiniband support enabled ED2 fails with a segmentation fault. Does anyone else use ED2 on servers with infiniband?  Are there additional compile options I need?  Error message:

```
[sserbin@modex build]$ ./ed_2.1-opt

Program received signal SIGSEGV: Segmentation fault - invalid memory reference.

Backtrace for this error:
#0  0x7FB468E58287
#1  0x7FB468E5888E
#2  0x309223269F
#3  0x7FB4681404F3
#4  0x7FB468141D52
#5  0x7FB46813F1CE
#6  0x7FB4681305A6
#7  0x7FB468126712
#8  0x7FB468126040
#9  0x7FB46812480D
#10  0x7FB468215DCE
#11  0x7FB468213CE8
#12  0x7FB4680CEEA3
#13  0x7FB4680EE6BA
#14  0x7FB4685A0CA4
#15  0x40557F in MAIN__ at edmain.F90:0
Segmentation fault

```

@rgknox @mpaiao @DanielNScott

**Comment by serbinsh** [comment]
Using openmpi 1.8.4 with infiniband support

```
---------------------------------------------------- /data/software/environment-modules/modulefiles/ ----------------------------------------------------
autoconf/2.69                 gfortran/4.8.4                json-c/0.11(default)          maui/3.3.1                    perl/5.22.1
cdo/1.6.8                     hdf4/4.2.11                   json-c/0.12                   nco/4.4.8                     proj/4.8.0
gcc/4.8.0                     hdf5/1.8.13                   json-c/main                   netcdf/4.3.3.1                proj/4.9.1
gcc/4.8.4(default)            hdf5/1.8.14                   laszip/2.2.0                  openmpi/1.6.4-gnu48           ruby/ruby-2.1.0
gdal/1.11.2                   intel/2013                    libkml/devel                  openmpi/1.6.4-intel
geos/3.4.2                    intel/2013_sp1(default)       libtiff/4.0.4                 openmpi/1.6.5-intel
geotiff/1.4.1                 jags/3.4.0                    libxml2/2.9.2                 openmpi/1.8.4-gnu484(default)
gfortran/4.8.0                jags/4.4.0                    m4/1.4.17                     openmpi/1.8.4-gnu484-ib

```

**Comment by DanielNScott** [comment]
I haven't used the infiniband portion of our cluster, sorry to say.

**Comment by serbinsh** [comment]
OK, thanks.  I am not trying to use the infiniband support, that is to run ED in MPI using IB, but instead just get it running normally but on a system that has IB running.  But for some reason once it is enabled on the server ED2 can compile but wont run.  If you turn it off it works fine.  Very frustrating.

**Comment by serbinsh** [comment]
@DanielNScott I just realized, you are saying you have the support but don't run ED over IB correct?  Was there anything special you had to set in opt to get it to compile and run on servers with IB, but again not actually trying to use ED with IB?

Thanks

**Comment by DanielNScott** [comment]
Hmm, well so the cluster at Harvard (Odyssey) has nodes which are part of an infiniband network and nodes which aren't (MGHPCC nodes are, as I understand it, if an anyone is interested). I have only compiled the model on non-infiniband nodes, and only run it on non-infiniband nodes as well.

Re. your issue, my suggestion would be unloading all of your modules, cleaning the build directory, loading only the minimum set of modules you need to build and run ED, and then seeing if it will compile and run properly. My guess is that you're having an inter-software compatability issue, potentially involving the built executable, linked libraries, and/or modules, though it's really just a guess, and a fairly broad one.

**Comment by DanielNScott** [comment]
Actually @serbinsh it looks like maybe this (from your module list) is the issue:

```
openmpi/1.8.4-gnu484(default)
openmpi/1.8.4-gnu484-ib
```
I assume gnu484-ib is infiniband-compatible and gnu484 is not?

**Comment by serbinsh** [comment]
@DanielNScott Actually, those are the avail not the loaded.  I loaded the version with IB support when trying the compile. I have even tried to enable the USE_COLLECTIVE_MPIO=1 #0 FLAG but that also didn't fix the SEGFAULT....still trying

**Comment by serbinsh** [comment]
@DanielNScott I was thinking that maybe I need to re-compile HDF5 to include parallel support or whatever enables the ability to use IB and then re-compile ED2....even though I am not trying to run ED in parallel, other than that enabled by PEcAn.

**Comment by serbinsh** [comment]
At a bit of a standstill on this now.   I have tried enabling support for parallel with HDF5 hoping that might be the fix.  But so far it hasn't fixed the segmentation fault error I am getting.  Trying now to dig into debugging:

```
Segmentation fault
[sserbin@modex build]$ gdb -ex run ./ed_2.1-opt
GNU gdb (GDB) Red Hat Enterprise Linux (7.2-83.el6)
Copyright (C) 2010 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>...
Reading symbols from /data/sserbin/Modeling/ED2.git/ED/build/ed_2.1-opt...(no debugging symbols found)...done.
Starting program: /data/sserbin/Modeling/ED2.git/ED/build/ed_2.1-opt
[Thread debugging using libthread_db enabled]
warning: File "/data/software/gcc-4.8.4/lib64/libstdc++.so.6.0.19-gdb.py" auto-loading has been declined by your `auto-load safe-path' set to "/usr/share/gdb/auto-load:/usr/lib/debug:/usr/bin/mono-gdb.py".
To enable execution of this file add
	add-auto-load-safe-path /data/software/gcc-4.8.4/lib64/libstdc++.so.6.0.19-gdb.py
line to your configuration file "/home/sserbin/.gdbinit".
To completely disable this security protection add
	set auto-load safe-path /
line to your configuration file "/home/sserbin/.gdbinit".
For more information about this security protection see the
"Auto-loading safe path" section in the GDB manual.  E.g., run from the shell:
	info "(gdb)Auto-loading safe path"
[New Thread 0x7ffff5b93700 (LWP 21169)]
[New Thread 0x7ffff4efe700 (LWP 21171)]

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff6af54f3 in udcm_module_finalize () from /data/software/openmpi/1.8.4-gnu484-ib/lib/libmpi.so.1
Missing separate debuginfos, use: debuginfo-install glibc-2.12-1.166.el6_7.7.x86_64 keyutils-libs-1.4-5.el6.x86_64 krb5-libs-1.10.3-42.el6.x86_64 libcom_err-1.41.12-22.el6.x86_64 libcxgb3-1.3.1-3.el6.x86_64 libibverbs-1.1.8-4.el6.x86_64 libipathverbs-1.3-3.el6_5.x86_64 libmlx4-1.0.6-7.el6.x86_64 libmlx5-1.0.2-1.el6.x86_64 libmthca-1.0.6-4.el6.x86_64 libnes-1.1.4-2.el6.x86_64 libnl-1.1.4-2.el6.x86_64 libselinux-2.0.94-5.8.el6.x86_64 libxml2-2.7.6-20.el6_7.1.x86_64 openssl-1.0.1e-42.el6_7.2.x86_64 zlib-1.2.3-29.el6.x86_64

```

This could be a clue

```
0x00007ffff6af54f3 in udcm_module_finalize () from /data/software/openmpi/1.8.4-gnu484-ib/lib/libmpi.so.1

```

**Comment by serbinsh** [comment]
Its possible it is a openMPI version problem.  I do think it has everything to do with openMPI and the openIB module as I get seg faults with simple mpi programs run through mpirun as well....so it isn't necessarily an ED issue, though another issue may come up given that we are running with IB.

**Comment by serbinsh** [comment]
@DanielNScott uncovered the issue we were having on Modex:

```
Hi Shawn, Toni,

I figured out what was causing the segfault errors once I enabled IB 
support on the modex cluster.  The issue was actually with the resource 
manager (torque pbs_mom) not having enough memory resources available 
due to a user max locked memory limit (this is a default).

I added the following line to /etc/sysconfig/pbs_mom:
ulimit -l unlimited

(Note that file was not there. I had to create this file)

Try your tests again and let me know how you make out.

btw- I changed to the Mellanox OFED (IB) drivers on all nodes. These 
drivers are pretty solid and newer than what's available from the 
distribution.  Let's just keep these running for now.
```

ED2 is now working OK.

```
[sserbin@modex build]$ ./ed_2.1-opt
+---------------- MPI parallel info: --------------------+
+  - Machnum  =      0
+  - Machsize =      1
+---------------- OMP parallel info: --------------------+
+  - thread  use:       1
+  - threads max:       1
+  - cpu     use:       1
+  - cpus    max:       1
+  Note: Max vals are for node, not sockets.
+--------------------------------------------------------+
Reading namelist information
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::

--------------------------------------------------------------
                     !!! FATAL ERROR !!!
--------------------------------------------------------------
    ---> File:        ed_load_namelist.f90
    ---> Subroutine:  read_nl
    ---> Reason:      The namelist file ED2IN is missing.
--------------------------------------------------------------
 ED execution halts (see previous error message)...
--------------------------------------------------------------
STOP fatal_error
```

So good news!
